packages/medusa-payment-paypal/src/services/paypal-provider.ts

```diff
@@ -0,0 +1,386 @@
+
+  /**
+   * Paypal does not provide such feature
+   * @param paymentSessionData
+   */
+  async deletePayment(
+    paymentSessionData: Record<string, unknown>
+  ): Promise<
+    PaymentProcessorError | PaymentProcessorSessionResponse["session_data"]
+  > {
+    return paymentSessionData
+  }
+
+  async refundPayment(
+    paymentSessionData: Record<string, unknown>,
+    refundAmount: number
+  ): Promise<
+    PaymentProcessorError | PaymentProcessorSessionResponse["session_data"]
+  > {
+    const { purchase_units } = paymentSessionData as {
+      purchase_units: PurchaseUnits
+    }
+
+    try {
+      const purchaseUnit = purchase_units[0]
+      const payments = purchaseUnit.payments
+      const isAlreadyCaptured = purchase_units.some(
+        (pu) => pu.payments.captures?.length
+      )
+
+      if (!isAlreadyCaptured) {
+        throw new Error("The order has not yet been captured. Unable to refund")
+      }
+
+      const paymentId = payments.captures[0].id
+      const currencyCode = purchaseUnit.amount.currency_code
+      const request = new PayPal.payments.CapturesRefundRequest(paymentId)
+
+      request.requestBody({
+        amount: {
+          currency_code: currencyCode,
+          value: roundToTwo(
+            humanizeAmount(refundAmount, currencyCode),
+            currencyCode
+          ),
+        },
+      })
+
+      await this.paypal_.execute(request)
+
+      return this.retrievePayment(paymentSessionData)
+    } catch (error) {
+      return this.buildError("An error occurred in refundPayment", error)
+    }
+  }
+
+  async retrievePayment(
+    paymentSessionData: Record<string, unknown>
+  ): Promise<
+    PaymentProcessorError | PaymentProcessorSessionResponse["session_data"]
+  > {
+    try {
+      const id = paymentSessionData.id as string
+      const request = new PayPal.orders.OrdersGetRequest(id)
+      const res = await this.paypal_.execute(request)
+      return res.result
+    } catch (e) {
+      return this.buildError("An error occurred in retrievePayment", e)
+    }
+  }
+
+  async updatePayment(
+    context: PaymentProcessorContext
+  ): Promise<PaymentProcessorError | PaymentProcessorSessionResponse | void> {
+    try {
+      const { currency_code, amount } = context
+      const id = context.paymentSessionData.id as string
+
+      const request = new PayPal.orders.OrdersPatchRequest(id)
+      request.requestBody([
+        {
+          op: "replace",
+          path: "/purchase_units/@reference_id=='default'",
+          value: {
+            amount: {
+              currency_code: currency_code.toUpperCase(),
+              value: roundToTwo(
+                humanizeAmount(amount, currency_code),
+                currency_code
+              ),
+            },
+          },
+        },
+      ])
+
+      await this.paypal_.execute(request)
+
+      return { session_data: context.paymentSessionData }
+    } catch (error) {
+      return this.initiatePayment(context).catch((e) => {
```

Comment:
todo: Add missing await
```suggestion
      return await this.initiatePayment(context).catch((e) => {
```